Configure a default logout procedure

bbuehrle opened the pull request:
First ever pull request so please go easy on me :)

I used your tutorial from [Medium](https://medium.com/@ro0NL/adding-user-management-to-your-symfony-application-ceeefe2a2e9) to VERY quickly start a base Symfony 4 project and add authentication from the get-go.  Being still fairly new to symfony, the only issue that I had was that the only way to logout (that I noticed) was in the debug toolbar.  I added a few extra lines, and I'm hoping for the best!

Any comments/suggestions on my first try are welcome.

bbuehrle committed:
Add default logout to main firewall

bbuehrle committed:
Add route to allow logout

bbuehrle committed:
Add link in profile to quickly logout

bbuehrle committed:
Add option to activate route for /logout

GawainLynch commented:
Welcome and congratulations on being brave enough to make your first PR! It gets much easier from here.  :beers:

bbuehrle committed:
Remove whitespace in blank line

ro0NL commented on src/UserBundle/Maker/UserMaker.php:
```diff
@@ -433,6 +433,10 @@ private function generateControllers(ConsoleStyle $io): void
             ])];
         }
 
+        if ($hasLogin && $io->confirm('Activate /logout route (Add config/routes.yaml)?')) {
```
we need to account for this question in generated security.yaml (in case the answer is "no" we should not add the logout section).

so lets ask this first, before we ask about security.yaml

ro0NL replied:
also i'd like to ask for `Can users logout? (adds config/routes.yaml)`

ro0NL commented on src/UserBundle/Resources/skeleton/routes.php:
```diff
@@ -0,0 +1,13 @@
+#    controller: App\Controller\DefaultController::index
+logout:
+    path: /logout
+
```
one newline to many :)

ro0NL reviewed (commented):
this is a missing piece indeed :+1:

bbuehrle committed:
Merge branch 'master' into bbuehrle-logout

bbuehrle committed:
Suggested changes
- Change wording of logout question
- Reorder logout question before generating security.yaml
- Remove extra new line in routes.php
- Add logic to dynamically add logout to security.yaml

Signed-off-by: Brian Buehrle <bbueh02@outlook.com>

bbuehrle committed:
Merge remote-tracking branch 'origin/bbuehrle-logout' into bbuehrle-logout

ro0NL commented on src/UserBundle/Maker/UserMaker.php:
```diff
@@ -426,10 +426,17 @@ private function generateControllers(ConsoleStyle $io): void
 PHP;
         }
 
+        $wantsLogout = $io->confirm('Can users logout? (adds config/routes.yaml)');
```
```
$hasLogout = $hasLogin && $io->confirm('...');
```

so it's only asked in case of login being available, whereas currently it's always asked but the actual write happens conditionally with `$hasLogin`

ro0NL commented on src/UserBundle/Maker/UserMaker.php:
```diff
@@ -426,10 +426,17 @@ private function generateControllers(ConsoleStyle $io): void
+            $this->writes[] = [$this->projectDir . '/config/packages/security.yaml', self::getSkeleton('security.php', [
                 'hashAlgorithm' => $this->getPassordHashAlgorithm(),
                 'fieldName' => $usernameField,
+                'logout' => $securityLogout,
```
```
'logout' => $hasLogout,
```

ro0NL commented on src/UserBundle/Resources/skeleton/security.php:
```diff
@@ -31,7 +31,7 @@
                 username_parameter: ${fieldName}
                 password_parameter: password
-
+${logout}
```
you can declare `$logoutSection` based on `if ($logout) ...` which we assigned above, that avoids the extra file

ro0NL reviewed (commented):
almost there :)

bbuehrle committed:
Refactor logout section

Signed-off-by: Brian Buehrle <bbueh02@outlook.com>

bbuehrle commented:
I've never used the HEREDOC before so I'm hoping that this is what you were referring to.  Let me know and I'll look into it further if I need to.

ro0NL commented:
@bbuehrle i tried to add a commit, but permission denied. Can you apply this diff, then we're done.

```diff
diff --git a/src/UserBundle/Resources/skeleton/security.php b/src/UserBundle/Resources/skeleton/security.php
index a2f9c1e..371d593 100644
--- a/src/UserBundle/Resources/skeleton/security.php
+++ b/src/UserBundle/Resources/skeleton/security.php
@@ -2,9 +2,11 @@
 
 declare(strict_types=1);
 
+$logoutSection = '';
+
 if ($logout) {
     $logoutSection = <<<YAML
-            
+
             logout:
                 path: /logout
                 target: /
diff --git a/src/UserBundle/Resources/skeleton/template/profile.html.php b/src/UserBundle/Resources/skeleton/template/profile.html.php
index a0f00bf..293b13f 100644
--- a/src/UserBundle/Resources/skeleton/template/profile.html.php
+++ b/src/UserBundle/Resources/skeleton/template/profile.html.php
@@ -5,11 +5,13 @@ declare(strict_types=1);
 use MsgPhp\UserBundle\Twig\GlobalVariable;
 
 $userVariable = GlobalVariable::NAME;
+$logoutSection = '';
 
 if ($logout) {
-    $logoutSection = <<<LOGOUT
+    $logoutSection = <<<HTML
+
     <p><a href="{{ path('logout') }}">Logout</a></p>
-LOGOUT;
+HTML;
 }
 
 return <<<TWIG
@@ -18,8 +20,7 @@ return <<<TWIG
 {% block ${block} %}
     <h1>Your Profile</h1>
 
-    <p>Logged in as: <em>{{ ${userVariable}.current.${fieldName} }}</em></p>
-    ${logoutSection}
+    <p>Logged in as: <em>{{ ${userVariable}.current.${fieldName} }}</em></p>${logoutSection}
 {% endblock %}
 
 TWIG;
```

Yes the HEREDOC is hard sometimes :")

ro0NL commented:
@bbuehrle wait :) i just realized... we actually don't need routing; there's `logout_path()` in twig: http://symfony.com/doc/current/reference/twig_reference.html#logout-path

IMHO we can enable this by default, so there's no need for the extra question during make:user. It's simply matter of adding `logout: true` to security.yaml and then use `{{ logout_path() }}` in the profile template.

ro0NL committed:
Merge branch 'master' into bbuehrle-logout

ro0NL committed:
patch

ro0NL commented:
seems that i do have permissions :)

thanks @bbuehrle

bbuehrle commented:
Sorry for not responding and finishing this all up.  For some reason, I stopped getting notifications about your comments.  Is this all taken care of then, or do you still need me to do the diff?  I could have sworn I allowed the permissions on the branch...

ro0NL commented:
@bbuehrle all good :) feel free to test the latest version (master branch) as i've not released this yet.
